Feature/embedded entities

pmattmann opened the pull request:
#1579 Embedded Entities Beispiel aufbauen

**Idee:**
Es werden pro Entität jeweils folgende Serialization-Groups angelegt:
- read
  Ist bei allen Lese-Operationen aktiv.
  Properties markiert mit **read** werden immer ausgegeben
- write
  Ist bei allen Schreib-Operationen aktiv.
  Properties markiert mit **write** sind immer schreibbar (post & patch)
- create
  ist bei Post-Operation aktiv
  Properties markeirt mit **create** sind nur beim erstellen schreibbar (post)
- update
  Ist bei Patch-Operation aktiv
  Properties markiert mit **update** sind änderbar (patch)

Zusätzlich werden Serialization-Groups pro Relation **Quelle:Ziel** erstellt.
Diese erlauben selektives embedden von solchen Relationen.

Damit die Relation selektiv embedded werden kann, wird jeweils das Property der Relation mit \#\[ApiProperty(readableLink:false)\] markiert.
Zusätzlich wird eine Getter-Methode **getEmbeddedXYZ()** erstellt. Diese wird mit \#\[ApiProperty(readableLink:true)\] und \#\[Groups(Quelle:Ziel)\] markiert.


Siehe Beispiele:
- read: [Day.dayOffset](https://github.com/ecamp/ecamp3/pull/1682/files#diff-aad2ad2f4ea183eef766802685a47b2a743737e1a3bf5dc156b3fdfe2a116596R67)
- write: [Camp.title](https://github.com/ecamp/ecamp3/pull/1682/files#diff-bcaffa096e98f188f76cfd0d9f00de45486bc22d3bfe4d89570b3e97ee1720b7R155)
- create: [Camp.name](https://github.com/ecamp/ecamp3/pull/1682/files#diff-bcaffa096e98f188f76cfd0d9f00de45486bc22d3bfe4d89570b3e97ee1720b7R142)
- linked Entity: [Period.Camp](https://github.com/ecamp/ecamp3/pull/1682/files#diff-b30b2806f6900f62e0fc4ed70c0b130723dd9d59d17d42ff14291d6737b19cf2R90)
- embedded Entity: [Period.Camp](https://github.com/ecamp/ecamp3/pull/1682/files#diff-b30b2806f6900f62e0fc4ed70c0b130723dd9d59d17d42ff14291d6737b19cf2R150)
- linked Collection: [Period.Days](https://github.com/ecamp/ecamp3/pull/1682/files#diff-b30b2806f6900f62e0fc4ed70c0b130723dd9d59d17d42ff14291d6737b19cf2R150)
- embedded Collection: [Period.Days](https://github.com/ecamp/ecamp3/pull/1682/files#diff-b30b2806f6900f62e0fc4ed70c0b130723dd9d59d17d42ff14291d6737b19cf2R150)


Durch angeben von Relationen **Quelle:Ziel** im **normalizationContext** kann bestimmt werden, ob eine Relation embedded werden soll, oder nicht.

pmattmann committed:
Embedded Collections camps/periods/days

pmattmann committed:
Merge remote-tracking branch 'ecamp/devel' into cosinus/api-validation

carlobeltrame commented on api/src/Entity/Camp.php:
```diff
@@ -25,20 +25,32 @@
             'security' => 'object.owner == user or is_granted("ROLE_ADMIN")',
             'denormalization_context' => [
-                'groups' => ['camp:update'],
+                'groups' => ['Properties:write', 'Camp:update'],
```
Müsste hier nicht auch noch `Camp:write` stehen? Sonst sind Relations die mit `Camp:write` markiert sind nicht durch PATCH schreibbar. So wie ich es verstehe sollte `Camp:write` ja einfach ein Superset von `Camp:create` und `Camp:update` sein, also sobald ich eine Property mit `Camp:write` markiere ist sie bei jeglichen Schreiboperationen auf `/camps/*` URLS schreibbar.

pmattmann replied:
done

carlobeltrame reviewed (changes requested):
Ich finde den Vorschlag grundsätzlich gut. Statt `Properties:read` / `Properties:write` könnte ich mir auch `Any:read` / `Any:write` vorstellen, da man es dann immer als eine Einschränkung nach Root Entity Klasse lesen kann. Aber zugegebenermassen wird das wohl fast immer nur für primitive Properties verwendet, daher wäre auch dein Vorschlag okay für mich.

Eine Limitierung dieses Systems mit den Groups tritt auf, wenn man rekursive Datenstrukturen hat. Z.B. wenn wir die `children` eines `ContentNode`s mit `writableLink: true` markieren würden, also nested POST / nested PATCH erlauben würden, dann könnte man nicht separate Regeln für Payload-Root und nested Children festlegen.
Ein anderes Beispiel wo das vielleicht schlimmer sein könnte, wäre wenn wir z.B. MaterialItem und MaterialList gegenseitig nested schreibbar machen. Wenn ich dann ein MaterialItem patche ist die Group `MaterialItem:write` aktiv. Im Payload könnte ich also nested die MaterialList und darin nested ein beliebiges anderes MaterialItem mitsenden und so mitbearbeiten.
Aber ich glaube solange wir `writableLink: true` nur sparsam und bedacht einsetzen, sollte das in der Praxis kein grosses Problem sein.

pmattmann committed:
Add Camp:write

pmattmann commented:
Allgemeine Notiz:
Mit der vorgeschlagenen Groups-Logik funktionieren die UnitTests aktuell nicht.
Grund ist die Idee **ECampApiTestCase::getExamplePayload**.

Diese versucht anhand einer API-Resource einen Beispiel-Payload automatisch zu erzeugen. Die Schema-Namen der Api-Resourcen ändern jedoch zusammen mit den Groups. Das führt dazu, dass diese Schema-Namen je nach refactoring immer wieder ändern können / nicht stabil sind.
Evtl. sind wir gut beraten selber Biespiel-Payloads pro HTTP-Verb und Entität zu erstellen.

-> Bitte diskutieren.

pmattmann committed:
selectively embed referenced entitiy

pmattmann committed:
excample fixed

carlobeltrame commented on api/src/Entity/Camp.php:
```diff
@@ -56,9 +74,13 @@ class Camp extends BaseEntity implements BelongsToCampInterface {
+    #[ApiProperty(
+        readableLink: false,
+        writableLink: true,
+        example: '/periods?camp=/camps/1a2b3c4d',
```
Hier hätte ich gerne mein altes Example zurück, sonst muss ich in Swagger sehr viel tippen um ein Camp zu erstellen

carlobeltrame commented on api/src/Entity/Camp.php:
```diff
@@ -25,20 +25,38 @@
+            'denormalization_context' => [
+                'groups' => ['write', 'create', 'Camp:Periods'],
+                'allow_extra_attributes' => false,
+            ],
```
Brauchts bei Create / Update auch noch einen `normalization_context`? Würde das überhaupt funktionieren?

pmattmann replied:
ja, das funktioniert 👍

carlobeltrame commented on api/src/Entity/Camp.php:
```diff
@@ -116,7 +138,7 @@ class Camp extends BaseEntity implements BelongsToCampInterface {
     #[Assert\NotBlank]
     #[ApiProperty(example: 'SoLa 2022')]
-    #[Groups(['Default', 'camp:update'])]
+    #[Groups(['read', 'create'])]
```
```suggestion
    #[Groups(['read', 'write'])]
```
Sonst kann man den Campnamen nie mehr updaten

carlobeltrame commented on api/src/Entity/Camp.php:
```diff
@@ -128,8 +150,8 @@ class Camp extends BaseEntity implements BelongsToCampInterface {
     #[Assert\Length(max: 32)]
-    #[ApiProperty(example: 'Abteilungs-Sommerlager 2022')]
-    #[Groups(['Default', 'camp:update'])]
+    #[ApiProperty(writable: true, example: 'Abteilungs-Sommerlager 2022')]
```
```suggestion
    #[ApiProperty(example: 'Abteilungs-Sommerlager 2022')]
```
`writable: true` ist der Default

carlobeltrame commented on api/src/Entity/Day.php:
```diff
@@ -44,7 +53,8 @@ class Day extends BaseEntity implements BelongsToCampInterface {
      * @ORM\JoinColumn(nullable=false, onDelete="cascade")
      */
-    #[ApiProperty(writable: false, example: '/periods/1a2b3c4d')]
+    #[ApiProperty(readableLink: false, writable: false, example: '/periods/1a2b3c4d')]
```
Ist `readableLink: false` nicht der Default?

pmattmann replied:
ne, bei Entitäten nicht; bei Collection glaub ja

carlobeltrame commented on api/src/Entity/Day.php:
```diff
@@ -44,7 +53,8 @@ class Day extends BaseEntity implements BelongsToCampInterface {
      * @ORM\ManyToOne(targetEntity="Period", inversedBy="days")
```
Fehlt bei `public Collection $dayResponsibles` noch ein `#[Groups(['read'])]`?

carlobeltrame commented on api/src/Entity/Day.php:
```diff
@@ -53,13 +63,15 @@ class Day extends BaseEntity implements BelongsToCampInterface {
     }
 
-    #[ApiProperty(readable: false)]
+    #[ApiProperty(readableLink: false, writable: false, example: '/camps/1a2b3c4d')]
```
Das Camp ist jetzt neu lesbar auf dem Day? Könnte man das rückgängig machen?

carlobeltrame commented on api/src/Entity/Period.php:
```diff
@@ -38,7 +52,12 @@ class Period extends BaseEntity implements BelongsToCampInterface {
      */
-    #[ApiProperty(writable: false, example: '["/days/1a2b3c4d"]')]
+    #[ApiProperty(
+        readableLink: false,
```
Ist `readableLink: false` nicht der Default?

carlobeltrame commented on api/src/Entity/Period.php:
```diff
@@ -66,8 +85,8 @@ class Period extends BaseEntity implements BelongsToCampInterface {
      */
-    #[ApiProperty(example: '/camps/1a2b3c4d')]
-    #[Groups(['Default'])]
+    #[ApiProperty(readableLink: false, example: '/camps/1a2b3c4d')]
```
Ist `readableLink: false` nicht der Default?

carlobeltrame commented on api/src/Entity/Period.php:
```diff
@@ -79,7 +98,7 @@ class Period extends BaseEntity implements BelongsToCampInterface {
     #[Assert\NotBlank]
     #[ApiProperty(example: 'Hauptlager')]
-    #[Groups(['Default', 'period:update'])]
+    #[Groups(['Properties:read', 'Properties:write'])]
```
Diese Gruppen heissen jetzt anders

carlobeltrame commented on api/src/Entity/Period.php:
```diff
@@ -95,7 +114,7 @@ class Period extends BaseEntity implements BelongsToCampInterface {
     #[Assert\LessThanOrEqual(propertyPath: 'end')]
     #[ApiProperty(example: '2022-01-01', openapiContext: ['format' => 'date'])]
-    #[Groups(['Default', 'period:update'])]
+    #[Groups(['Properties:read', 'Properties:write'])]
```
Diese Gruppen heissen jetzt anders

carlobeltrame commented on api/src/Entity/Period.php:
```diff
@@ -106,7 +125,7 @@ class Period extends BaseEntity implements BelongsToCampInterface {
     #[Assert\GreaterThanOrEqual(propertyPath: 'start')]
     #[ApiProperty(example: '2022-01-08', openapiContext: ['format' => 'date'])]
-    #[Groups(['Default', 'period:update'])]
+    #[Groups(['Properties:read', 'Properties:write'])]
```
Diese Gruppen heissen jetzt anders

carlobeltrame commented on api/src/Entity/Period.php:
```diff
@@ -115,6 +134,23 @@ public function __construct() {
+        return $this->days;
+    }
+
+    #[ApiProperty(readableLink: true)]
```
Bei `getEmbeddedDays` hast du ein Example drin, hier nicht. Müsste man das ergänzen oder fällt er zurück auf das Example der Property?

carlobeltrame replied:
Bei den `getEmbedded*` Methoden brauchts kein Example, das wird sowieso nicht in Swagger angezeigt. Stattdessen werden, wenn man den Return Type korrekt annotiert, die eingebetteten Felder perfekt automatisch angezeigt, inklusive den korrekten Normalisierungs-Gruppen :partying_face: 
```php
    /**
     * @return Day[]
     */
    #[ApiProperty(readableLink: true)]
    #[SerializedName('days')]
    #[Groups('Period:Days')]
    public function getEmbeddedDays(): array {
        return $this->days->getValues();
    }

    #[ApiProperty(readableLink: true)]
    #[SerializedName('camp')]
    #[Groups(['Period:Camp'])]
    public function getEmbeddedCamp(): ?Camp {
        return $this->camp;
    }
```
führt zu folgendem Output in Swagger:
![Screenshot from 2021-08-04 10-13-22](https://user-images.githubusercontent.com/7566995/128146439-a5218641-ebc2-4c83-88e9-88bebee58dfe.png)

pmattmann committed:
fix post on camp

pmattmann committed:
code-review changes

carlobeltrame committed:
Sort named arguments

carlobeltrame committed:
Make the camp name writable again

carlobeltrame committed:
Make dayResponsibles readable again

carlobeltrame committed:
Revert readability of camp on day

carlobeltrame committed:
Fix return type declarations to make examples in swagger work automatically

carlobeltrame committed:
getExamplePayload needs to take into account different schema for different operations

carlobeltrame committed:
Always add some defaults to the context

https://symfonycasts.com/screencast/api-platform-security/service-decoration
https://api-platform.com/docs/core/serialization/#changing-the-serialization-context-dynamically

carlobeltrame committed:
Add missing normalization groups so the tests pass again

carlobeltrame committed:
Work around broken circular reference detection in HAL format

See https://github.com/api-platform/core/issues/4358

carlobeltrame committed:
Ignore the embedded getters when trying to read the Doctrine metadata

carlobeltrame committed:
Consistently use our group name system for all entities

Also, deactivate the automatic embedding feature of API platform,
because with our serialization groups system, everything would always
get embedded everywhere. We want to only activate embedding selectively.

carlobeltrame committed:
Temporarily replace API platform's eager loading extension

This is until https://github.com/api-platform/core/pull/4377 is merged,
see that PR and https://github.com/api-platform/core/issues/2444 for
details.

carlobeltrame committed:
Merge branch 'devel' into feature/embedded-entities

carlobeltrame reviewed (approved):
Ich habe die Tests gefixt und die vorgeschlagenen Gruppen bei allen Entities eingefügt.
Danach hat API Platform automatisch mal sämtliche Relationen embedded, auch alle die wir nicht so markiert haben. Bei der Fehlersuche bin ich auf ein verstecktes Stück [Dokumentation](https://github.com/api-platform/docs/pull/1403/files) gestossen, und mir ist einiges in Bezug auf Embedding in API Platform klarer geworden.

Automatisch embedded (d.h. automatisch `readableLink=true`) wird eine Relation dann, wenn 1. die Relation selber durch die aktiven Gruppen lesbar ist und 2. zumindest ein Feld in der related Entity durch die aktiven Gruppen lesbar ist. Normalerweise schlägt API Platform vor (implizit in den Docs und explizit in den SymfonyCasts), Serialisierungsgruppen ala `camp:update`, `period:read` zu verwenden. Bei uns ist aber der Vorschlag, `update`, `read` etc. zu verwenden. Dadurch sind beide obigen Konditionen gegeben, und alles wird überall automatisch embedded.

Ich finde unseren Ansatz trotzdem noch spannend wegen der Eleganz mit der `Camp:Periods` Config die er für unseren Fall bietet. Um das so zum fliegen zu bringen, musste ich daher ein paar Anpassungen an API Platform vornehmen. Eine davon ist in api-platform/core#4377 bei API Platform angebracht (aber bei uns vorgeholt), die anderen (CircularReferenceDetectingHalItemNormalizer, PreventAutomaticEmbeddingPropertyMetadataFactory) sind spezifisch für unseren Fall, die habe ich nur bei uns implementiert.

To discuss: Halten wir an unserem Vorschlag für Serializer Groups fest und unterhalten diese Abweichungen selber, oder gehen wir zurück zum inoffiziellen Standard?

BacLuc commented on api/src/Serializer/PreventAutomaticEmbeddingPropertyMetadataFactory.php:
```diff
@@ -0,0 +1,83 @@
+ * 30, so this class should be assigned a priority of 29.
+ * https://github.com/api-platform/core/blob/main/src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml#L65
+ */
+final class PreventAutomaticEmbeddingPropertyMetadataFactory implements PropertyMetadataFactoryInterface {
```
Bräuchten wir das  nicht, wenn wir vor read etc noch die den Namen der Entity packen würden?

carlobeltrame replied:
Korrekt. Es ist etwas kompliziert, aber der inoffizielle Standard für die Gruppen ist eben, `Period:read` etc. zu verwenden. Dann klappts aber mit dem Embedden nicht mehr ganz so elegant: Um im Camp die Perioden und die Days zu embedden müssten wir dann auf dem Camp den `normalization_context` mit den Gruppen `['Camp:read', 'Period:read', 'Day:read']` einrichten, statt wie vorgeschlagen `['read', 'Camp:Periods', 'Period:Days']`. Die Groups würden also Entities entsprechen statt Relations.

Einen Unterschied macht das, wenn man über verschiedene Relations zur gleichen Entity kommt: Bei ContentNodes könnte man mit den inoffiziell-Standard-Gruppen nur Root, Parent, Children und Descendants gleichzeitig embedden, oder alle nicht embedden. Z.B. nur den Root embedden geht nicht. Da müssten wir dann mit `#[MaxDepth(3)]` oder so arbeiten.

Dafür wäre es beim semi-Standard möglich, im embeddeten Objekt nur einen Teil der Properties auszugeben. Das wäre bei uns aber höchstens in Ausnahmefällen nützlich, etwa wenn wir im Profile Endpoint eine andere Repräsentation des Users embedden wollen als wenn man den User direkt via `/users/1a2b3c4d` abruft.

pmattmann committed:
Const for NormalizationCtx; use SwaggerDefName

carlobeltrame commented on api/src/Entity/Camp.php:
```diff
@@ -27,25 +27,30 @@
             'validation_groups' => ['Default', 'create'],
             'denormalization_context' => ['groups' => ['write', 'create']],
-            'normalization_context' => ['groups' => ['read', 'Camp:Periods', 'Period:Days']],
+            'normalization_context' => Camp::ITEM_NORMALIZATION_CONTEXT,
```
Wenn es ITEM_NORMALIZATION_CONTEXT heisst, ist es dann korrekt das auch für den collection normalization context zu verwenden?

pmattmann replied:
Naja - die Server-Antwort auf einen POST-Request auf den Collection-Endpoint ist ein einzelnes Item; keine Collection.
Daher finde ich - ja.

carlobeltrame replied:
Aha, clever, hast recht

BacLuc replied:
Um das copy pasta leben einfacher zu machen, könnte man statt
Camp::ITEM_NORMALIZATION_CONTEXT
self::ITEM_NORMALIZATION_CONTEXT
verwenden.

pmattmann replied:
done

pmattmann committed:
Merge remote-tracking branch 'ecamp/devel' into feature/embedded-entities

pmattmann committed:
prevent adding same alias twice; doctrine does not like it

pmattmann committed:
use self for ITEM_NORMALIZATION_CONTEXT

carlobeltrame committed:
Merge remote-tracking branch 'origin/devel' into feature/embedded-entities

carlobeltrame committed:
refactor: Extract method for preventing duplicate selects

carlobeltrame reviewed (approved):
Wäre jetzt nie darauf gekommen, aber b0cb899808be4d3047e249e8b4fb9190c1636e47 hat tatsächlich die Tests gefixt. Merci @pmattmann

pmattmann commented:
> Wäre jetzt nie darauf gekommen, aber [b0cb899](https://github.com/ecamp/ecamp3/commit/b0cb899808be4d3047e249e8b4fb9190c1636e47) hat tatsächlich die Tests gefixt. Merci @pmattmann

Jooo... war auch nicht einfach, den zu finden... musste ich ordentlich tief debuggen gehen...

BacLuc commented:
Die Coverage geht leider um 1.1 Prozent runter für api.
Wir haben ein threshold von 1% für den decrease in einem PR.
Gibt es noch Code, den man sinnvoll testen könnte, oder soll jemand (z.b. ich) den threshold auf 1.1 Prozent decrease stellen?

carlobeltrame commented:
Die eagerloadingextension ist wohl dafür verantwortlich. Die ist eigentlich nur temporär bei uns drin, bis api-platform/core#4377 gemerged und released ist. Wir könnten höchstens tests dafür noch bei uns reinmachen.

carlobeltrame committed:
Comment out unused code to prevent lowering the code coverage too much

The EagerLoadingExtension is only temporarily in our codebase, until
https://github.com/api-platform/core/pull/4377 is merged and released.

There are tests for this extension in API Platform core, so we shouldn't
need to write our own test suite for it. However, because of our limited
use cases so far, this class lowers the coverage too much. Therefore, I
commented out some code of which I am certain we won't need it in the
near future. Less untested code = higher coverage.

BacLuc commented on api/tests/Serializer/SerializerContextBuilderTest.php:
```diff
@@ -0,0 +1,79 @@
+
+        $result = $this->contextBuilder->createFromRequest($request, false);
+
+        $this->assertNotEquals(['skip_null_values' => false], $result);
```
minor:
Hätte ich jetzt eher positiv formuliert.
Falls möglich möchte ich wissen, was zurückgegeben wird. Nicht dass das $result nicht diesem Array entspricht.

BacLuc commented on api/tests/Serializer/SerializerContextBuilderTest.php:
```diff
@@ -0,0 +1,79 @@
+
+        $result = $this->contextBuilder->createFromRequest($request, true);
+
+        $this->assertNotEquals(['allow_extra_attributes' => false], $result);
```
minor:
Hätte ich jetzt eher positiv formuliert.
Falls möglich möchte ich wissen, was zurückgegeben wird. Nicht dass das $result nicht diesem Array entspricht.
